### ISSUE TITLE
Type checking for a tuple of objects returns an error

### ISSUE THREAD

**Issue by matheusfelipeog** [microsoft/pyright#6355]
## Problem

When I define an index on a variable or a class attribute to access a specific object in a tuple, I get an error.

For example, the code:

```python
class T1:
    idx = 0


class T2:
    idx = 1


t1_idx = 0
t2_idx = 1


class Foo:
    def __init__(self) -> None:
        self.data = (T1(), T2())

    def get_t1(self) -> T1:
        return self.data[T1.idx]

    def get_t2(self) -> T2:
        return self.data[t2_idx]
```

Returns the following error:

```
Expression of type "T1 | T2" cannot be assigned to return type "T1"
    Type "T1 | T2" cannot be assigned to type "T1"
      "T2" is incompatible with "T1" (reportGeneralTypeIssues)

Expression of type "T1 | T2" cannot be assigned to return type "T2"
    Type "T1 | T2" cannot be assigned to type "T2"
      "T1" is incompatible with "T2" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
```

But if I use a literal index directly when accessing the tuple's value:

```python
class Foo:
    def __init__(self) -> None:
        self.data = (T1(), T2())

    def get_t1(self) -> T1:
        return self.data[0]

    def get_t2(self) -> T2:
        idx = 1
        return self.data[idx]
```

I don't get an error:

```
0 errors, 0 warnings, 0 informations
```

Am I missing something, or is it indeed a bug?

## Env

This specific test was executed in the CLI, but the error occurs both when using the CLI and the VSCode extension in their latest versions.

- Python 3.11.6
- pyright 1.1.334

You can also check the code in the [Pyright playground](https://pyright-play.net/?code=MYGwhgzhAEAqCMAuAUNN0CWATAHtAvNAAzKmiQywBMK6muB08pyALvAPrZ6EmtVcGhZmXBRoAMQD2U2uiwBTAGbQOXAHYZWagBQQFIJQEpoAWgB80AHJT1CuXTT7DAOixhWYRjoQ6jAGjgqPyNSOkUVAHMFbXY9A2MzSwQHRwAnGIBXNPVoZyU3DzAAbQQXbgBdMPllaGjY4PyTCyDUugzWbNz8ws9i-kEcKtEKSRkaVBqVNQxNbQ54w2bLGzs29B73T29fAKCQ6rQIupiOOKakuCRJx2gOrryE3pKiYfDa%2BrPGhOXWm8duIxmLc0Pcco9XFsSpVkEA).

**Comment by erictraut** [comment]
This behavior is correct. You're using a variable to index into the tuple, so the type checker cannot assume which element you're going to access. The value of `T1.idx`, for example, could be reassigned to `1`.

You can fix this in your code by doing one of the following:

1. Use all-caps names (e.g. change `idx` to `IDX`). This tells pyright to treat the symbol as a constant.
2. Use a `Final` type annotation (e.g. `idx: Final = 0`). This will likewise tell pyright to treat the symbol as a constant.
3. Use a `Literal` annotation (e.g. idx: Literal[0] = 0). In this case, the symbol remains a variable, but it can be assigned only a value of 0, so pyright can now safely assume that its value will not change.
